input/s3/s3_input.go

```diff
@@ -0,0 +1,133 @@
+package s3
+
+import (
+	"archive/tar"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"hash"
+	"io"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/rlmcpherson/s3gof3r"
+	"github.com/spacemonkeygo/errors"
+	"github.com/spacemonkeygo/errors/try"
+	"polydawn.net/repeatr/def"
+	"polydawn.net/repeatr/input"
+	"polydawn.net/repeatr/input/tar2"
+	"polydawn.net/repeatr/lib/fshash"
+)
+
+const Type = "s3"
+
+var _ input.Input = &Input{} // interface assertion
+
+type Input struct {
+	spec          def.Input
+	hasherFactory func() hash.Hash
+}
+
+func New(spec def.Input) input.Input {
+	if spec.Type != Type {
+		panic(errors.ProgrammerError.New("This input implementation supports definitions of type %q, not %q", Type, spec.Type))
+	}
+	return &Input{
+		spec:          spec,
+		hasherFactory: sha512.New384,
+	}
+}
+
+func (i Input) Apply(destinationRoot string) <-chan error {
+	done := make(chan error)
+	go func() {
+		defer close(done)
+		try.Do(func() {
+			// do make a dir for untaring into.
+			// tars may specify permission and time bits for their top dir, but if not, we'll start with sane defaults.
+			err := os.MkdirAll(destinationRoot, 0755)
+			if err != nil {
+				panic(input.TargetFilesystemUnavailableIOError(err))
+			}
+
+			// parse URI
+			u, err := url.Parse(i.spec.URI)
+			if err != nil {
+				panic(input.ConfigError.New("failed to parse URI: %s", err))
+			}
+			bucketName := u.Host
+			storePath := u.Path
+			var splay bool
+			switch u.Scheme {
+			case "s3":
+				splay = false
+			case "s3+splay":
+				splay = true
+			default:
+				panic(input.ConfigError.New("unrecognized scheme: %q", u.Scheme))
+			}
+
+			// load keys from env
+			// TODO someday URIs should grow smart enough to control this in a more general fashion -- but for now, host ENV is actually pretty feasible and plays easily with others.
+			keys, err := s3gof3r.EnvKeys()
+			if err != nil {
+				panic(S3CredentialsMissingError.Wrap(err))
+			}
+
+			// initialize reader from s3!
+			getPath := storePath
+			if splay {
+				getPath = path.Join(storePath, i.spec.Hash)
+			}
+			s3reader := makeS3reader(bucketName, getPath, keys)
+			defer s3reader.Close()
+
+			// prepare decompression as necessary
+			reader, err := tar2.Decompress(s3reader)
+			if err != nil {
+				panic(input.DataSourceUnavailableError.New("could not start decompressing: %s", err))
+			}
+			tarReader := tar.NewReader(reader)
+
+			// unroll the tar, copying and accumulating data for integrity check
+			bucket := &fshash.MemoryBucket{}
+			tar2.Extract(tarReader, destinationRoot, bucket, i.hasherFactory)
+
+			// hash whole tree
+			actualTreeHash, _ := fshash.Hash(bucket, i.hasherFactory)
+
+			// verify total integrity
+			expectedTreeHash, err := base64.URLEncoding.DecodeString(i.spec.Hash)
+			if !bytes.Equal(actualTreeHash, expectedTreeHash) {
+				done <- input.InputHashMismatchError.New("expected hash %q, got %q", i.spec.Hash, base64.URLEncoding.EncodeToString(actualTreeHash))
```

Comment:
Am I correct in understanding that, to use this implementation, you're required to:

1. Only use s3 assets produced (or scanned) by repeatr's custom sha384 logic
1. Preserve the custom sha384 along with the upload URL, and not the s3-specific hash (md5)

To use this transport?

For example, it sounds like the workflow to use s3 assets right now would be:

1. Download the s3 asset with some other tool (probably the `gof3r` binary, actually)
1. Scan it with repeatr
1. Provide it to a formula, probably as a `dir` input

Both limitations should be fine for a demonstration; I want to make sure I fully understand.